pkg/work/spoke/apply/server_side_apply_test.go

```diff
@@ -279,3 +294,251 @@ func newManifestWork() *unstructured.Unstructured {
+				"spec": map[string]interface{}{
+					"name": "name1",
+				},
+			}},
+			jsonPaths: []string{".spec.replicas"},
+		},
+		{
+			name: "remove multiple fields",
+			req: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"name":     "name1",
+					"replicas": int64(1),
+					"containers": []interface{}{
+						map[string]interface{}{
+							"image": "test",
+						},
+					},
+				},
+			}},
+			exp: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"name": "name1",
+				},
+			}},
+			jsonPaths: []string{".spec.replicas", ".spec.containers"},
+		},
+		{
+			name: "remove filtered fields",
+			req: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"name":     "name1",
+					"replicas": int64(1),
+					"containers": []interface{}{
+						map[string]interface{}{
+							"name":  "container1",
+							"image": "test",
+						},
+						map[string]interface{}{
+							"name":  "container2",
+							"image": "test",
+						},
+					},
+				},
+			}},
+			exp: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"replicas": int64(1),
+					"name":     "name1",
+					"containers": []interface{}{
+						map[string]interface{}{
+							"name": "container1",
+						},
+						map[string]interface{}{
+							"name":  "container2",
+							"image": "test",
+						},
+					},
+				},
+			}},
+			jsonPaths: []string{".spec.containers[?(@.name==\"container1\")].image"},
+		},
+		{
+			name: "list field is kept",
+			req: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"name":     "name1",
+					"replicas": int64(1),
+					"containers": []interface{}{
+						map[string]interface{}{
+							"name":  "container1",
+							"image": "test",
+						},
+						map[string]interface{}{
+							"name":  "container2",
+							"image": "test",
+						},
+					},
+				},
+			}},
+			exp: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"replicas": int64(1),
+					"name":     "name1",
+					"containers": []interface{}{
+						map[string]interface{}{
+							"name":  "container1",
+							"image": "test",
+						},
+						map[string]interface{}{
+							"name":  "container2",
+							"image": "test",
+						},
+					},
+				},
+			}},
+			jsonPaths: []string{".spec.containers[?(@.name==\"container1\")]"},
+		},
+	}
+	logger := klog.NewKlogr()
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			desired := c.req.DeepCopy()
+			for _, jsonPath := range c.jsonPaths {
+				removeFieldByJSONPath(desired.UnstructuredContent(), jsonPath, logger)
+			}
+			if !equality.Semantic.DeepEqual(c.exp, desired) {
+				t.Errorf("expected %v, got %v", c.exp, desired)
+			}
+		})
+	}
+}
+
+func TestServerSideApplyWithIgnoreFields(t *testing.T) {
```

Comment:
Does it make sense if ignorefields is .metadata.xxx? For example, what if ignorefields is .metadata.annotations and type is IgnoreOnSpokePresent, will the existing annotations be kept on the spoke or be override by object-hash?

Reply:
it will be kept. We compute the hash for the whole object including metadata.